File: 2022/2022W07/PreppindataW07.py
import pandas as pd
import re

people_xlsx = pd.ExcelFile(r"C:\Data\PreppinData\PeopleData.xlsx")
data_xlsx = pd.ExcelFile(r"C:\Data\PreppinData\MetricData2021.xlsx")

location = pd.read_excel(people_xlsx,'Location').set_index('Location ID').T.to_dict('records')[0]

agent = pd.read_excel(people_xlsx,'People')
agent['Agent Name'] = agent['last_name'] + ', ' + agent['first_name']
agent['Location'] = agent['Location ID'].apply(lambda x: location[x])

leader = pd.read_excel(people_xlsx,'Leaders')
leader['Leader Name'] = leader['last_name'] + ', ' + leader['first_name']

agent = pd.merge(agent, leader[['id', 'Leader Name']].rename(columns={'id':'Leader 1'}), on='Leader 1')

date_dim = pd.read_excel(people_xlsx,'Date Dim')
date_dim['Year'] = date_dim['Month Start Date'].dt.year
date_dim['Month Start Date'] = date_dim['Month Start Date'].dt.date
date_dim = date_dim[date_dim['Year']==2021][['Month Start Date']]

correct_cols = {'AgentID':'id', 'Offered':'Calls Offered', 'Not Answered':'Calls Not Answered', 'Answered':'Calls Answered'}
temp = []
for sheet in data_xlsx.sheet_names:
    df = pd.read_excel(data_xlsx,sheet).rename(columns=correct_cols)
    df.insert(0,'Month Start Date',pd.to_datetime(f'2021-{sheet}-01', format='%Y-%b-%d'))
    temp.append(df)
call_data = pd.concat(temp)   
call_data['Month Start Date'] = call_data['Month Start Date'].dt.date

final = pd.merge(agent, date_dim, how='cross')
final = pd.merge(final, call_data, on=['id','Month Start Date'], how='left')  
goal = pd.read_excel(people_xlsx,'Goals')['Goals'].tolist()
for g in goal:
    final[g] = int(re.search('(\d+)$', g).group(1))
final['Not Answered Rate'] = round(final['Calls Not Answered']/final['Calls Offered'],3)
final['Agent Avg Duration'] = round(final['Total Duration']/final['Calls Answered'])
final['Met Not Answered Rate'] = (final['Not Answered Rate']*100 < final['Not Answered Percent < 5'])
final['Met Sentiment Goal'] = (final['Sentiment'] >= final['Sentiment Score >= 0'])

final = final[['id', 'Agent Name', 'Leader 1', 'Leader Name', 'Month Start Date','Location', 'Calls Answered', 'Calls Not Answered', 
               'Not Answered Rate', 'Met Not Answered Rate', 'Not Answered Percent < 5', 'Calls Offered', 'Total Duration', 
               'Agent Avg Duration', 'Transfers', 'Sentiment', 'Sentiment Score >= 0', 'Met Sentiment Goal']]

File: 2022/2022W07/preppindataW07.R
library(readxl)
library(dplyr)
library(tidyr)
library(stringr)
library(lubridate)
library(purrr)

Sys.setlocale("LC_ALL","English")
people_xlsx <- "C:/Data/PreppinData/PeopleData.xlsx"
data_xlsx <- "C:/Data/PreppinData/MetricData2021.xlsx"

correct_cols <- c('id'='AgentID', 'Calls Offered'='Offered', 'Calls Not Answered'='Not Answered', 'Calls Answered'='Answered')

call_data <- data_xlsx %>%
  excel_sheets() %>%
  set_names() %>%
  map_df(
    ~ read_excel(path = data_xlsx, sheet = .x,) %>% rename(any_of(correct_cols)),
    .id = 'Month') %>%
  mutate('Month Start Date' = as.Date(paste0('2021-',`Month`,'-01'), format='%Y-%b-%d'))

final <- read_excel(people_xlsx, 'People') %>%
  mutate('Agent Name' = paste(`last_name`, `first_name`, sep=', ')) %>%
  merge(read_excel(people_xlsx, 'Location'), by = 'Location ID') %>%
  merge(read_excel(people_xlsx, 'Leaders') %>%
          mutate('Leader Name' = paste(`last_name`, `first_name`, sep=', '),
                 'Leader 1' = `id`) %>%
          select(c('Leader 1', 'Leader Name')), by='Leader 1') %>%
  crossing(., read_excel(people_xlsx, 'Date Dim') %>% filter(year(`Month Start Date`) == 2021)) %>%
  crossing(., read_excel(people_xlsx, 'Goals') %>%
             mutate('value' = as.numeric(str_extract(`Goals`, '(\\d+)$'))) %>%
             pivot_wider(names_from = 'Goals', values_from = 'value')) %>%
  merge(call_data, by=c('id', 'Month Start Date'), all.x=TRUE) %>%
  mutate('Not Answered Rate' = round(`Calls Not Answered`/`Calls Offered`,3),
         'Agent Avg Duration' = round(`Total Duration`/`Calls Answered`),
         'Met Not Answered Rate' = (`Not Answered Rate`*100 < `Not Answered Percent < 5`),
         'Met Sentiment Goal' = (`Sentiment` >= `Sentiment Score >= 0`)) %>%
  select(c('id', 'Agent Name', 'Leader 1', 'Leader Name', 'Month Start Date','Location', 'Calls Answered', 'Calls Not Answered', 
           'Not Answered Rate', 'Met Not Answered Rate', 'Not Answered Percent < 5', 'Calls Offered', 'Total Duration', 
           'Agent Avg Duration', 'Transfers', 'Sentiment', 'Sentiment Score >= 0', 'Met Sentiment Goal'))

View(final)
